Opentelemetry tutorial revise (#22648)
backstage/backstage: docs/tutorials/setup-opentelemetry.md

Author: Yes. I didn't manage to get it to work with imports. But `require` should work just fine. There is an example I added in the tutorial

## Configure

> In your `packages/backend/src` folder, create an `instrumentation.ts` file.

Author: Not sure why it has to be outside /src, but Docker can't find it otherwise. Maybe someone could explain this

Reviewer: @SonilPro The .dockerignore file ignores packages/*/src.  An exclusion for the file in the dockerignore could be added using !packages/backend/src/instrumentation.ts

Author: Ahh okay. That makes sense. I can move the file to source then if it makes sense, but we could also keep it outside.

> In your `packages/backend` folder, create an `instrumentation.ts` file.

Reviewer 2: I am not sure that this should be a .ts file, right? Sounds to me like this should be js

Author: Yeah, .js should be enough

```typescript
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { ConsoleSpanExporter } = require('@opentelemetry/sdk-trace-node');
const {
  getNodeAutoInstrumentations,
} = require('@opentelemetry/auto-instrumentations-node');
const {
  PeriodicExportingMetricReader,
  ConsoleMetricExporter,

> } = require('@opentelemetry/sdk-metrics');

Reviewer 3: these are missing typehints now I believe :/ Is there a way to use require and maintain typehints?

Author: Honestly, I am not sure. I am not that good at typescript/javascript. I'd like to get this PR merged, maybe someone in the future will have a better idea for this whole setup, but for now I think it will do just fine. ESLint still works, so that is a plus

sdk.start();
```

You probably won't need all of the instrumentation inside `getNodeAutoInstrumentations()` so make sure to
check the [documentation](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node) and tweak it properly.

It's important to setup the NodeSDK and the automatic instrumentation **before** importing any library.

This is why we will use the nodejs [`--require`](https://nodejs.org/api/cli.html#-r---require-module)
flag when we start up the application.

> In your `Dockerfile` add the `--require` flag which points to the `instrumentation.ts` file

Author: ```suggestion
In your `Dockerfile` add the `--require` flag which points to the `instrumentation.js` file
```

```Dockerfile

# We need the instrumentation file inside the Docker image so we can use it with --require
// highlight-add-next-line
COPY --chown=node:node packages/backend/instrumentation.js ./

// highlight-remove-next-line
CMD ["node", "packages/backend", "--config", "app-config.yaml"]
// highlight-add-next-line

> CMD ["node", "--require", "./instrumentation.ts", "packages/backend", "--config", "app-config.yaml"]

Author: ```suggestion
CMD ["node", "--require", "./instrumentation.js", "packages/backend", "--config", "app-config.yaml"]
```

```

## Run Backstage

The above configuration will only work in production once your start a Docker container from the image.

> To be able to test locally you can import the `./instrumentation.ts` file at the top (before all imports) of your backend `index.ts` file

Author: ```suggestion
To be able to test locally you can import the `./instrumentation.js` file at the top (before all imports) of your backend `index.ts` file
```

sdk.start();
```

> Your probably won't need all the instrumentations inside `getNodeAutoInstrumentations()` so make sure to

Reviewer 4: ```suggestion
Your probably won't need all the instrumentation inside `getNodeAutoInstrumentations()` so make sure to
```

check the [documentation](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node) and tweak it properly.

It's important to setup the NodeSDK and the automatic instrumentation **before** importing any library.

This is why we will use the nodejs [`--require`](https://nodejs.org/api/cli.html#-r---require-module)
flag when we start up the application.

In your `Dockerfile` add the `--require` flag which points to the `instrumentation.ts` file

```Dockerfile
FROM node:18-bookworm-slim

> ...

Reviewer 4: Would there be other content here? I find this can confuse people so maybe we just leave a comment to that fact

// highlight-add-next-line
COPY --chown=node:node packages/backend/instrumentation.ts ./

COPY --chown=node:node yarn.lock package.json packages/backend/dist/skeleton.tar.gz ./
RUN tar xzf skeleton.tar.gz && rm skeleton.tar.gz

RUN --mount=type=cache,target=/home/node/.yarn/berry/cache,sharing=locked,uid=1000,gid=1000 \
    yarn workspaces focus --all --production

COPY --chown=node:node packages/backend/dist/bundle.tar.gz app-config*.yaml ./
RUN tar xzf bundle.tar.gz && rm bundle.tar.gz

// highlight-remove-next-line
CMD ["node", "packages/backend", "--config", "app-config.yaml"]
// highlight-add-next-line
CMD ["node", "packages/backend", "--require", "./instrumentation.ts" "--config", "app-config.yaml"]
```

## Run Backstage

You can now start your Backstage instance as usual, using `yarn dev`.

When the backend is started, you should see in your console traces and metrics emitted by OpenTelemetry.

Of course in production you probably won't use the console exporters but instead send traces and metrics to an OpenTelemetry Collector or other exporter using [OTLP exporters](https://opentelemetry.io/docs/instrumentation/js/exporters/).

> If you need to disable/configure some Opentelemetry feature there are lots of [environment variables](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/) which you can tweak.

Reviewer 4: ```suggestion
If you need to disable/configure some OpenTelemetry feature there are lots of [environment variables](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/) which you can tweak.
```

sdk.start();
```

> Your probably won't need all the instrumentation inside `getNodeAutoInstrumentations()` so make sure to

Reviewer 2: ```suggestion
You probably won't need all of the instrumentation inside `getNodeAutoInstrumentations()` so make sure to
```

check the [documentation](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node) and tweak it properly.

It's important to setup the NodeSDK and the automatic instrumentation **before** importing any library.

This is why we will use the nodejs [`--require`](https://nodejs.org/api/cli.html#-r---require-module)
flag when we start up the application.

In your `Dockerfile` add the `--require` flag which points to the `instrumentation.ts` file

```Dockerfile
FROM node:18-bookworm-slim
...
# More functionality goes here
...
WORKDIR /app
RUN chown node:node /app
USER node

ENV NODE_ENV production

COPY --chown=node:node .yarn ./.yarn
COPY --chown=node:node .yarnrc.yml ./

# We need the instrumentation file inside the Docker image so we can use it with --require
// highlight-add-next-line

> COPY --chown=node:node packages/backend/instrumentation.ts ./

Reviewer 3: This line should probably go right before the CMD, so that the docker caching works better, right? Also backstage should probably add COPY --link to this file. Its part of the best practices for dockerfiles

https://docs.docker.com/reference/dockerfile/#copy---link

Author: For this I just followed the current state of the Dockerfile. Maybe @freben knows more about this. For now I changed the Dockerfile so it will be up to the user to fit the commands where he wants.

COPY --chown=node:node yarn.lock package.json packages/backend/dist/skeleton.tar.gz ./
RUN tar xzf skeleton.tar.gz && rm skeleton.tar.gz

RUN --mount=type=cache,target=/home/node/.yarn/berry/cache,sharing=locked,uid=1000,gid=1000 \
    yarn workspaces focus --all --production

COPY --chown=node:node packages/backend/dist/bundle.tar.gz app-config*.yaml ./
RUN tar xzf bundle.tar.gz && rm bundle.tar.gz

// highlight-remove-next-line
CMD ["node", "packages/backend", "--config", "app-config.yaml"]
// highlight-add-next-line

> CMD ["node", "--require", "./instrumentation.ts", "packages/backend", "--config", "app-config.yaml"]

Reviewer 2: I'd say remove everything above except this removal and addition. It's going to quickly diverge from reality and how the actual Dockerfile looks otherwise, over time.

> CMD ["node", "--require", "./instrumentation.ts", "packages/backend", "--config", "app-config.yaml"]

Reviewer 3: Is this needed? When running it locally, loading the `instrumentation.ts` at the top of my index.ts seemed to work.

Author: The whole reason for this revise is because it doesn't perfectly when deployed. Yes, adding the instrumentation file will work locally, but once you deploy, the code will get bundled and that will break some of the opentelemetry functionalities. For the second comment, check that you don't have some .dockerignore filter blocking this

Reviewer 3: Hmm my docker ignore looks _ok_, but I can see that the file is not preset in the docker container, so it must not be copied over for some reason 😕 
```
node_modules
packages/*/src
packages/*/node_modules
plugins
*.local.yaml
```

shouldnt match `packages/backend/instrumentation.ts`

Reviewer 3: Yeah I mean looking at the dockerfile, i see no reason why this would be copied into the container

Reviewer 3: Oh man, i missed this line 
```
COPY --chown=node:node packages/backend/instrumentation.ts ./
```
sorry about that!

```

## Run Backstage

The above configuration will only work in production once your start a Docker container from the image.

> To be able to test locally you can import the `./instrumentation.ts` file at the top (before all imports) of your backend `index.ts` file

Reviewer 3: ```suggestion
To be able to test locally you can import the `../instrumentation.ts` file at the top (before all imports) of your backend `index.ts` file
```

```ts

> import '../instrumentation.ts'

Reviewer 3: I had a hard time getting prettier to stop moving this lower. My (bad) workaround was to add this `packages/backend/src/index.ts` to prettier ignore

Author: @fordneild  Hmm. Didn't have this problem. Maybe you have some custom prettier configuration? But yes, adding it there will fix that.
